quick_sort.py

```diff
@@ -0,0 +1,62 @@
+def quick_sort(array):
+    '''Input an array, the lowest value, and the highest value. Perform a
+    quicksort and return a sorted list'''
+    return _quick_sort_helper(array, 0, (len(array) - 1))
+
+
+def _quick_sort_helper(array, lo, hi):
+    if lo < hi:
+        partition = _partition(array, lo, hi)
+        _quick_sort_helper(array, lo, (partition - 1))
+        _quick_sort_helper(array, (partition + 1), hi)
+    return array
+
+
+def _partition(array, lo, hi):
+    pivot_index = ((lo + hi)/2)
+    pivot_value = array[pivot_index]
+    array[pivot_index], array[hi] = array[hi], array[pivot_index]
+    stored_index = lo
+    for i in range(lo, hi + 1):
+        if array[i] < pivot_value:
+            array[i], array[stored_index] = array[stored_index], array[i]
+            stored_index += 1
+    array[stored_index], array[hi] = array[hi], array[stored_index]
+    return stored_index
+
+
+if __name__ == "__main__":
+    from time import time
+
+    l = range(100000)
+
+    timer = time()
+    quick_sort(l)
+    print "best case: "+str(time() - timer)
+
+    def shitty_quick_sort(array):
+        return _quick_sort_helper(array, 0, (len(array) - 1))
+
+    def shitty_quick_sort_helper(array, lo, hi):
+        if lo < hi:
+            partition = _partition(array, lo, hi)
+            _quick_sort_helper(array, lo, (partition - 1))
+            _quick_sort_helper(array, (partition + 1), hi)
+        return array
+
+    def shitty_partition(array, lo, hi):
+        """same as non-shitty partition, but selects lowest index"""
+        pivot_index = hi
+        pivot_value = array[pivot_index]
+        array[pivot_index], array[hi] = array[hi], array[pivot_index]
+        stored_index = lo
+        for i in range(lo, hi + 1):
+            if array[i] < pivot_value:
+                array[i], array[stored_index] = array[stored_index], array[i]
+                stored_index += 1
+        array[stored_index], array[hi] = array[hi], array[stored_index]
+        return stored_index
```

Comment:
It is quite odd to have to re-produce the entire stack of methods from your good qsort just to replace the thing that picks the pivot index.  Maybe you could make that somehow independent of the sort so that you can substitute it more easily?  This is a pretty smelly way to to the job.  (plus I'm pretty sure that you are using the good partition method anyway).